subsys/bluetooth/host/conn.c

```diff
@@ -614,429 +592,460 @@ static inline uint16_t conn_mtu(struct bt_conn *conn)
+		(conn->has_data == NULL);
 }
 
-static int conn_prepare_events(struct bt_conn *conn,
-			       struct k_poll_event *events)
+struct bt_conn *get_conn_ready(void)
 {
-	if (!atomic_get(&conn->ref)) {
-		return -ENOTCONN;
-	}
+	/* Here we only peek: we pop the conn (and insert it at the back if it
+	 * still has data) after the QoS function returns false.
+	 */
+	sys_snode_t *node  = sys_slist_peek_head(&bt_dev.le.conn_ready);
 
-	if (conn->state == BT_CONN_DISCONNECTED &&
-	    atomic_test_and_clear_bit(conn->flags, BT_CONN_CLEANUP)) {
-		conn_cleanup(conn);
-		return -ENOTCONN;
+	if (node == NULL) {
+		return NULL;
 	}
 
-	if (conn->state != BT_CONN_CONNECTED) {
-		return -ENOTCONN;
+	struct bt_conn *conn = CONTAINER_OF(node, struct bt_conn, _conn_ready);
+
+	if (cannot_send_to_controller(conn)) {
+		/* We will get scheduled again when the buffers are freed. */
+		LOG_DBG("no LL bufs for %p", conn);
+		return NULL;
 	}
 
-	LOG_DBG("Adding conn %p to poll list", conn);
+	if (dont_have_tx_context(conn)) {
+		/* We will get scheduled again when TX contexts are available. */
+		LOG_DBG("no TX contexts");
+		return NULL;
+	}
 
-	/* ISO Synchronized Receiver only builds do not transmit and hence
-	 * may not have any tx buffers allocated in a Controller.
-	 */
-	struct k_sem *conn_pkts = bt_conn_get_pkts(conn);
+	CHECKIF(dont_have_methods(conn)) {
+		LOG_DBG("conn %p (type %d) is missing mandatory methods",
+			conn, conn->type);
 
-	if (!conn_pkts) {
-		return -ENOTCONN;
+		return NULL;
 	}
 
-	bool buffers_available = k_sem_count_get(conn_pkts) > 0;
-	bool packets_waiting = !k_fifo_is_empty(&conn->tx_queue);
+	/* this also works for ISO */
+	if (bt_data_should_stop(conn)) {
+		sys_snode_t *s = sys_slist_get(&bt_dev.le.conn_ready);
+
+		__ASSERT_NO_MSG(s == node);
+		(void)s;
 
-	if (packets_waiting && !buffers_available) {
-		/* Only resume sending when the controller has buffer space
-		 * available for this connection.
+		atomic_t old = atomic_set(&conn->_conn_ready_lock, 0);
+		/* Note: we can't assert `old` is non-NULL here, as the
+		 * connection might have been marked ready by an l2cap channel
+		 * that cancelled its request to send.
 		 */
-		LOG_DBG("wait on ctlr buffers");
-		k_poll_event_init(&events[0],
-				  K_POLL_TYPE_SEM_AVAILABLE,
-				  K_POLL_MODE_NOTIFY_ONLY,
-				  conn_pkts);
-	} else {
-		/* Wait until there is more data to send. */
-		LOG_DBG("wait on host fifo");
-		k_poll_event_init(&events[0],
-				  K_POLL_TYPE_FIFO_DATA_AVAILABLE,
-				  K_POLL_MODE_NOTIFY_ONLY,
-				  &conn->tx_queue);
+
+		(void)old;
+
+		/* Append connection to list if it still has data */
+		if (conn->has_data(conn)) {
+			LOG_DBG("appending %p to back of TX queue", conn);
+			bt_conn_data_ready(conn);
+		}
 	}
-	events[0].tag = BT_EVENT_CONN_TX_QUEUE;
 
-	return 0;
+	return conn;
 }
 
-int bt_conn_prepare_events(struct k_poll_event events[])
+/* Crazy that this file is compiled even if this is not true, but here we are. */
```

Comment:
Sorry :(

Reply:
We can maybe alleviate this by moving all the TX stuff from `conn.c` to `conn_tx.c` or something like that, so that `conn.c` is exclusively for connected connections and not unconnected connections like BIS